utils/fetch_external_bangs.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python
+
+import sys
+import json
+import re
+from os.path import realpath, dirname, join
+
+import requests
+
+# set path
+sys.path.append(realpath(dirname(realpath(__file__)) + '/../'))
+
+from searx import searx_dir
+
+
+# from https://duckduckgo.com/newbang
+URL_BV1 = 'https://duckduckgo.com/bv1.js'
+
+RE_BANG_VERSION = re.compile('\/bang\.v([0-9]+)\.js')
+
+
+def get_bang_url():
+    response = requests.get(URL_BV1)
+    response.raise_for_status()
+    
+    r = RE_BANG_VERSION.findall(response.text)
+    return f'https://duckduckgo.com/bang.v{r[0]}.js', r[0]
+
+
+def fetch_bangs(url):
+    response = requests.get(url)
+    response.raise_for_status()
+    return json.loads(response.content.decode())
+
+
+def merge_when_no_leaf(node):
+    """Minimize the number of nodes
+
+    A -> B -> C 
+    B is child of A
+    C is child of B
+
+    If there C equals to '*', then each C are merged into A
+
+    For example:
+      d -> d -> g -> * (ddg*)
+        -> i -> g -> * (dig*)
+    becomes
+      d -> dg -> *
+        -> ig -> *
+    """
+    restart = False
+    if not isinstance(node, dict):
+        return
+    keys = list(node.keys())
+
+    for key in keys:
+        if key == '*':
+            continue
+
+        value = node[key]
+        value_keys = list(value.keys())
+        if '*' not in value_keys:
+            for value_key in value_keys:
+                node[key + value_key] = value[value_key]
+                merge_when_no_leaf(node[key + value_key])
+            del node[key]
+            restart = True
+        else:
+            merge_when_no_leaf(value)
+
+    if restart:
+        merge_when_no_leaf(node)
+
+
+def optimize_leaf(parent, parent_key, node):
+    if not isinstance(node, dict):
+        return
+
+    if len(node) == 1 and '*' in node and parent is not None:
+        parent[parent_key] = node['*']
+    else:
+        for key, value in node.items():
+            optimize_leaf(node, key, value)
+
+
+def parse_bangs(bangs):
+    bang_trie = {}
+    bang_urls = {}
+
+    for bang_definition in bangs:
+        # bang_trie
+        bang = bang_definition['t']
+
+        # bang_list
+        bang_url = bang_definition['u']
+        if '{{{s}}}' not in bang_url:
```

Comment:
Nit: you can get rid of the comment by refactoring the code

```python

def _is_invalid_bang_url(bang_url):
    return '{{{s}}}' not in bang_url
```
then call it here:
```python
if _is_invalid_bang_url(bang_url):
    continue
```

Reply:
For me the existing implementation is more readable .. and function calls are (minimal) more time consuming .. but I would have another nit :-) ..  I would place cancellation criteria at the beginning  (checking `bang_url` before eval `bang` in line 93).

Reply:
I move the code at the top of the loop, so the pattern is obvious